pkg/tbtc/maintainer/bitcoin_difficulty.go

```diff
@@ -2,54 +2,271 @@ package maintainer
+	// error logs in case of a permanent error in the Bitcoin difficulty
+	// maintainer.
+	defaultRestartBackoffTime = 120 * time.Second
+
+	// Default value for backoff time which should be applied after each attempt
+	// to prove a single Bitcoin epoch.
+	defaultEpochProvenBackOffTime = 60 * time.Second
+
+	// The number of blocks in a Bitcoin epoch.
+	bitcoinEpochLength = 2016
+)
+
 var logger = log.Logger("keep-maintainer-bitcoin-difficulty")
 
-func newBitcoinDifficultyMaintainer(
+func initializeBitcoinDifficultyMaintainer(
 	ctx context.Context,
 	btcChain bitcoin.Chain,
 	chain BitcoinDifficultyChain,
-) *BitcoinDifficultyMaintainer {
+	epochProvenBackOffTime time.Duration,
+	restartBackOffTime time.Duration,
+) {
 	bitcoinDifficultyMaintainer := &BitcoinDifficultyMaintainer{
-		btcChain: btcChain,
-		chain:    chain,
+		btcChain:               btcChain,
+		chain:                  chain,
+		epochProvenBackOffTime: epochProvenBackOffTime,
+		restartBackOffTime:     restartBackOffTime,
 	}
 
 	go bitcoinDifficultyMaintainer.startControlLoop(ctx)
-
-	return bitcoinDifficultyMaintainer
 }
 
 // BitcoinDifficultyMaintainer is the part of maintainer responsible for
 // maintaining the state of the Bitcoin difficulty on-chain contract.
 type BitcoinDifficultyMaintainer struct {
 	btcChain bitcoin.Chain
 	chain    BitcoinDifficultyChain
+
+	epochProvenBackOffTime time.Duration
+	restartBackOffTime     time.Duration
 }
 
-// startControlLoop launches the loop responsible for controlling the Bitcoin
+// startControlLoop starts the loop responsible for controlling the Bitcoin
 // difficulty maintainer.
-func (r *BitcoinDifficultyMaintainer) startControlLoop(ctx context.Context) {
+func (bdm *BitcoinDifficultyMaintainer) startControlLoop(ctx context.Context) {
 	logger.Info("starting Bitcoin difficulty maintainer")
 
 	defer func() {
 		logger.Info("stopping Bitcoin difficulty maintainer")
 	}()
 
 	for {
+		err := bdm.proveEpochs(ctx)
+		if err != nil {
+			logger.Errorf(
+				"restarting Bitcoin difficulty maintainer due to error while "+
+					"proving Bitcoin blockchain epochs [%v]",
+				err,
+			)
+		}
+
 		select {
+		case <-time.After(bdm.restartBackOffTime):
 		case <-ctx.Done():
 			return
-		default:
-			// TODO: Implement the maintainer loop. For now just print a message.
-			logger.Info("Bitcoin difficulty maintainer is working")
 		}
+	}
+}
 
-		time.Sleep(1 * time.Second)
+// proveEpochs proves Bitcoin blockchain epochs in the Bitcoin difficulty chain.
+func (bdm *BitcoinDifficultyMaintainer) proveEpochs(ctx context.Context) error {
+	if err := bdm.verifySubmissionEligibility(); err != nil {
+		return fmt.Errorf(
+			"cannot proceed with proving Bitcoin blockchain epochs [%v]",
+			err,
+		)
 	}
+
+	for {
+		if err := bdm.proveNextEpoch(); err != nil {
+			return fmt.Errorf("cannot prove Bitcoin blockchain epoch [%v]", err)
+		}
+
+		select {
+		case <-time.After(bdm.epochProvenBackOffTime):
```

Comment:
What is the reasoning behind using the backoff here? As I understand, we can have a couple of situations:
1. `bdm.proveNextEpoch()` returns an error, right now we return the error, so `epochProvenBackOffTime` won't be used.
2.  `bdm.proveNextEpoch()` succeeds but the relay contract is much behind the bitcoin chain, so a next epoch could be submitted to the relay right away.
3. `bdm.proveNextEpoch()` succeeds for the most recent epoch, and in this case, it's true that we should delay the next loop iteration.

Reply:
We have to wait here, because we need some time before the `LightRelay` smart contract is updated (the transaction has to be mined).
If we didn't wait here, the contract's `CurrentEpoch()` would return the same number and we would attempt to submit the same headers as before.

Reply:
I modified it a little bit in 42f53097e9843d1b5f3ab61fc7e34e70b1c01ae7.
Now we only use `epochProvenBackOffTime` after Bitcoin difficulty maintainer couldn't prove next epoch and we should wait.
I also renamed it to `idleBackOffTime`.